File: SEPTEMBER-2025/17-09-2025/p2353.cpp
struct comp {
    bool operator()(const pair<int,string> &a, const pair<int,string> &b) const {
        if (a.first == b.first) {
            return a.second < b.second;
        }
        return a.first > b.first;
    }
};
class FoodRatings {
public:
    unordered_map<string,set<pair<int,string>,comp>> mpp;
    unordered_map<string,pair<string,int>> foodMapping;
    FoodRatings(vector<string>& foods, vector<string>& cuisines, vector<int>& ratings) {
        int n = foods.size();
        for(int i = 0;i<n;i++){
            mpp[cuisines[i]].insert({ratings[i],foods[i]});
            foodMapping[foods[i]] = {cuisines[i],ratings[i]};
        }
    }
    
    void changeRating(string food, int newRating) {
        auto key = foodMapping[food];
        mpp[key.first].erase({key.second,food});
        mpp[key.first].insert({newRating,food});
        foodMapping.erase(food);
        foodMapping[food] = {key.first,newRating};
    }
    
    string highestRated(string cuisine) {
        auto &it = mpp[cuisine];

        string ans;
        for(auto st:it){
            ans = st.second;
            break;
        }
        return ans;
    }
};

/**
 * Your FoodRatings object will be instantiated and called as such:
 * FoodRatings* obj = new FoodRatings(foods, cuisines, ratings);
 * obj->changeRating(food,newRating);
 * string param_2 = obj->highestRated(cuisine);
 */

File: SEPTEMBER-2025/17-09-2025/p2353.py
import heapq

class FoodRatings:

    def __init__(self, foods: List[str], cuisines: List[str], ratings: List[int]):
        self.food_info = {}  # food -> (cuisine, rating)
        self.cuisine_map = {} # cuisine -> heap of (-rating, food)
        
        n = len(foods)
        for i in range(n):
            food = foods[i]
            cuisine = cuisines[i]
            rating = ratings[i]
            self.food_info[food] = (cuisine, rating)
            if cuisine not in self.cuisine_map:
                self.cuisine_map[cuisine] = []
            heapq.heappush(self.cuisine_map[cuisine], (-rating, food))
        

    def changeRating(self, food: str, newRating: int) -> None:
        cuisine, oldRating = self.food_info[food]
        self.food_info[food] = (cuisine, newRating)
        heapq.heappush(self.cuisine_map[cuisine], (-newRating, food))
        

    def highestRated(self, cuisine: str) -> str:
        heap = self.cuisine_map[cuisine]
        while heap:
            neg_rating, food = heap[0]
            current_rating = self.food_info[food][1]
            if -neg_rating != current_rating:
                heapq.heappop(heap)
            else:
                return food
        return ""  # should not happen if cuisine exists